can/zynqmp-can/zynqmp-can-test.c

```diff
@@ -0,0 +1,398 @@
+	int ret = zynqmp_can_send(oid, frames, 32, false, &sendFrames);
+	if (EOK == ret) {
+		printf("zynqmp-can-test: Successfully sent %u CAN frames\n", sendFrames);
+	}
+	else {
+		printf("zynqmp-can-test: Failed to send CAN frames, error %i\n", ret);
+		return ret;
+	}
+
+	/* Send frames again using the whole buffer; it should fail to send all frames */
+	ret = zynqmp_can_send(oid, frames, ARRAY_LENGTH(frames), false, &sendFrames);
+	if (ret == -ENOMEM) {
+		printf("zynqmp-can-test: As expected, failed to send all frames, sent only %u frames\n", sendFrames);
+	}
+	else {
+		printf("zynqmp-can-test: Unexpected test result, error code %i, sent %u frames\n", ret, sendFrames);
+	}
+
+	return ret;
+}
+
+/**
+ * Receives CAN frames from the RX FIFO once.
+ */
+static int recv_one_time(oid_t *oid)
+{
+	/* Read all frames from RX FIFO */
+	zynqmp_canFrame buf[64] = { 0 };
+	uint32_t recvFrames = 0;
+	time_t time = 0; /* Wait indefinitely */
+	int ret = zynqmp_can_recv(oid, buf, ARRAY_LENGTH(buf), time, &recvFrames);
+	if (ret == 0) {
+		printf("zynqmp-can-test: Successfully read %u CAN frames\n", recvFrames);
+		for (uint32_t i = 0; ((i < ARRAY_LENGTH(buf)) && (i < recvFrames)); i++) {
+			printf("CAN frame ID: 0x%x DLC: %u Data: ", buf[i].id, buf[i].len);
+			for (int n = 0; n < buf[i].len; n++) {
+				printf("0x%02X ", buf[i].payload.bytes[n]);
+			}
+			printf("\n");
+		}
+	}
+	else {
+		printf("zynqmp-can-test: Failed to receive frames, error %i\n", ret);
+	}
+
+	return ret;
+}
+
+/**
+ * Receives CAN frames from the RX FIFO with a timeout.
+ */
+static int recv_one_time_with_timeout(oid_t *oid)
+{
+	/* Read all frames from RX FIFO */
+	zynqmp_canFrame buf[64] = { 0 };
+	uint32_t recvFrames = 0;
+	time_t time = 5 * 1000 * 1000; /* Wait a few seconds */
+	int ret = zynqmp_can_recv(oid, buf, ARRAY_LENGTH(buf), time, &recvFrames);
+	if (ret == 0) {
+		printf("zynqmp-can-test: Successfully read %u CAN frames\n", recvFrames);
+		for (uint32_t i = 0; ((i < ARRAY_LENGTH(buf)) && (i < recvFrames)); i++) {
+			printf("CAN frame ID: 0x%x DLC: %u Data: ", buf[i].id, buf[i].len);
+			for (int n = 0; n < buf[i].len; n++) {
+				printf("0x%02X ", buf[i].payload.bytes[n]);
+			}
+			printf("\n");
+		}
+	}
+	else if (ret == -ETIME) {
+		printf("zynqmp-can-test: Received zero frames, timeout\n");
+	}
+	else {
+		printf("zynqmp-can-test: Failed to receive frames, error %i\n", ret);
+	}
+
+	return ret;
+}
+
+/**
+ * Enables or disables canonical mode and echo for the terminal.
+ */
+static void switchmode(int canon)
+{
+	struct termios state;
+
+	tcgetattr(STDIN_FILENO, &state);
+	if (canon) {
+		state.c_lflag |= ICANON;
+		state.c_lflag |= ECHO;
+	}
+	else {
+		state.c_lflag &= ~ICANON;
+		state.c_lflag &= ~ECHO;
+		state.c_cc[VMIN] = 1;
+	}
+	tcsetattr(STDIN_FILENO, TCSANOW, &state);
+}
+
+/**
+ * Receives CAN frames in a loop with a timeout.
+ * Displays received frame statistics in real time.
+ */
+static int recv_in_loop(oid_t *oid)
+{
+	switchmode(0);
```

Comment:
Standard idiom is to rather store initial termios state in a safe place and restore it at the end